base/gmp.jl

```diff
@@ -682,8 +682,16 @@ end
 
 factorial(x::BigInt) = isneg(x) ? BigInt(0) : MPZ.fac_ui(x)
 
+function binomial(n::BigInt, k::Integer)
+    k < 0 && return BigInt(0)
+    k <= typemax(UInt) && return binomial(n, UInt(k))
+    n < 0 && return isodd(k) ? -binomial(k - n - 1, k) : binomial(k - n - 1, k)
+    κ = n - k
+    κ < 0 && return BigInt(0)
+    κ <= typemax(UInt) && return binomial(n, UInt(κ))
+    throw(OverflowError("Computation will exceed memory"))
```

Comment:
Throwing an `OverflowError` here has the following effect on 64-bit Windows: For `κ > typemax(UInt64)`, an `OverflowError` is thrown, but for `typemax(UInt32) < κ ≤ typemax(UInt64)`, an `InexactError` is thrown.

This is because the C function that is ultimately called requires a `Culong`, which is equal to `UInt32` on Windows (but equal to `UInt` on other systems):
https://github.com/JuliaLang/julia/blob/a2ffdc904382971dcac3d3670f47ff06b9d127ac/base/gmp.jl#L171

I would prefer it if the type of error doesn’t depend on how big `κ` actually is. My suggestion would be to remove the explicit `throw`:
```suggestion
    return binomial(n, UInt(κ))
```
One could also replace every `UInt` in this function by `Culong` and replace `binomial(n::BigInt, k::UInt) = MPZ.bin_ui(n, k)` by `binomial(n::BigInt, k::Culong) = MPZ.bin_ui(n, k)`. This would avoid the unnecessary conversion to `UInt64` on 64-bit Windows.

Reply:
Thanks! Addressed the issue using the `Culong` solution so a less ambiguous throw is used.